GVFS/GVFS.Common/Tracing/TelemetryDaemonEventListener.cs

```diff
@@ -0,0 +1,162 @@
+            };
+
+            return JsonConvert.SerializeObject(message);
+        }
+
+        protected override void RecordMessageInternal(
+            string eventName,
+            Guid activityId,
+            Guid parentActivityId,
+            EventLevel level,
+            Keywords keywords,
+            EventOpcode opcode,
+            string payload)
+        {
+            string message = CreateJsonMessage(
+                this.vfsVersion,
+                this.providerName,
+                this.enlistmentId,
+                this.mountId,
+                eventName,
+                activityId,
+                parentActivityId,
+                level,
+                keywords,
+                opcode,
+                payload);
+
+            this.SendMessage(message);
+        }
+
+        private static string GetConfigValue(string gitBinRoot, string configKey)
+        {
+            string value = string.Empty;
+            string error;
+
+            GitProcess.ConfigResult result = GitProcess.GetFromSystemConfig(gitBinRoot, configKey);
+            if (!result.TryParseAsString(out value, out error, defaultValue: string.Empty) || string.IsNullOrWhiteSpace(value))
+            {
+                result = GitProcess.GetFromGlobalConfig(gitBinRoot, configKey);
+                result.TryParseAsString(out value, out error, defaultValue: string.Empty);
+            }
+
+            return value.TrimEnd('\r', '\n');
+        }
+
+        private void SendMessage(string message)
+        {
+            // Create pipe if this is the first message, or if the last connection broke for any reason
+            if (this.pipeClient == null)
+            {
+                string pipeName = GVFSPlatform.Instance.GetTelemetryNamedPipeName();
+                var pipe = new NamedPipeClientStream(".", pipeName, PipeDirection.Out, PipeOptions.Asynchronous);
+                try
+                {
+                    // Specify a instantaneous timeout because we don't want to hold up the rest of the
+                    // application if the pipe is not available; we will just drop this event.
+                    pipe.Connect(timeout: 0);
+                    this.pipeClient = pipe;
+                }
+                catch (TimeoutException)
+                {
+                    // We can't connect; we will try again with a new pipe on the next message
+                    return;
+                }
+            }
+
+            try
+            {
+                var buffer = Encoding.UTF8.GetBytes(message);
+                this.pipeClient.Write(buffer, 0, buffer.Length);
+            }
+            catch (Exception)
+            {
+                // We can't send this message for some reason (e.g., broken pipe); we attempt no recovery or retry
+                // mechanism and drop this message. We will try to recreate/connect the pipe on the next message.
+                this.pipeClient = null;

```

Comment:
do we need to dispose of pipeClient before nulling it it ?

Reply:
That's a good point. Most likely.. although I'm not sure what disposing of a broken pipe would do 🤔

Reply:
I've added a call to `Dispose` here too.